### PR TITLE
update how annotations are accessed

### PR DESCRIPTION
**Pull request by GiaJordan**
This PR updates how annotations from synapse are accessed by schematic. In `SynapaseStorage.format_row_annotations` we switched from using `syn.get_annotations` to using `SynapseStorage.get_async_annotations`. The latter gets the annotations from the API and the former gets them from the client.

Using the client annotations are formated:
`{"ID":[], <file annotations as key value pairs>}`
Using the API the annotations are formatted:
`{'annotations':{'id': value,... 'annotations': {<file annotations as key value pairs>}}}`

Schematic has been looking for annotations in the first level of the dict, this PR updates it to look for them in the nested dict where they are located now

----

Jenny edits:

I've wrapped up this PR with some minor changes:

    Merge with the new develop
    Added a new manifest file input - filepath_submission_test_manifest_sampleidx10.csv - which annotates the same synIDs but changes the SampleID annotation. This serves the purpose of ensuring that the annotations we see on Synapse are indeed the annotations that were uploaded during the current test run, and not a relic of a previous successful run.
    Re-introduce validating the uploaded annotations with the one from the mock_manifests manifest CSV
    Updated the Sample ID column to SampleID. The extra space was preventing being able to call the column by its name using the dataframe object, so it was being called with e.g. df._2. Now we can call with df.SampleID
    Implemented asyncio logic to retrieve the annotations from Synapse concurrently. This adds code complexity without improving the duration of the test by more than 2 seconds, so I lean towards removing it before merge, but I'm open to discussion.
    (out of scope) - upload-artifacts@v2 is no longer supported so I updated to v4 (more info [here](https://github.blog/changelog/2024-02-13-deprecation-notice-v1-and-v2-of-the-artifact-actions/))
    (out of scope) - The logic that caches the virtual environment was removed from test.yml

### CODE REVIEW

File: tests/integration/test_metadata_model.py
Comment (jaymedina) [reply tests/integration/test_metadata_model.py]: Hi @BryanFauble thanks for the suggestion, I went ahead and implemented it [here](https://github.com/Sage-Bionetworks/schematic/pull/1485/commits/6d76a93478e6b12e50a3f422e6593251fd5de501), however I found that it doesn't speed up the test by much - at most it seems to have shaved off 2 seconds from the 26 second duration. It seems the bottleneck is at the upload part which cannot be multi-threaded.

With the amount of complexity this new logic adds to the test, I don't think it's worth it to keep this as an async test, but let me know what you think.

**Comment by jaymedina** [comment]
I've wrapped up this PR with some minor changes:

1. Merge with the new `develop`
2. Added a new manifest file input - `filepath_submission_test_manifest_sampleidx10.csv` - which annotates the same synIDs but changes the `SampleID` annotation. This serves the purpose of ensuring that the annotations we see on Synapse are indeed the annotations that were uploaded during the current test run, and not a relic of a previous successful run.
3. Re-introduce validating the uploaded annotations with the one from the `mock_manifests` manifest CSV
4. Updated the `Sample ID` column to `SampleID`. The extra space was preventing being able to call the column by its name using the dataframe object, so it was being called with e.g. `df._2`. Now we can call with `df.SampleID`
5. Implemented `asyncio` logic to retrieve the annotations from Synapse concurrently. This adds code complexity without improving the duration of the test by more than 2 seconds, **so I lean towards removing it before merge, but I'm open to discussion**.

File: tests/data/mock_manifests/filepath_submission_test_manifest.csv
Comment (thomasyu888) [inline tests/data/mock_manifests/filepath_submission_test_manifest.csv:1]: May I ask why the change to `SampleID`?

File: tests/integration/test_metadata_model.py
Comment (thomasyu888) [inline tests/integration/test_metadata_model.py:101]: Nit: Should we move this import statement up top?

File: tests/data/mock_manifests/filepath_submission_test_manifest.csv
Comment (jaymedina) [reply tests/data/mock_manifests/filepath_submission_test_manifest.csv:1]: I updated the Sample ID column to SampleID because extra space was preventing being able to call the column by its name using the dataframe object, so it was being called with e.g. df._2. Now we can call with df.SampleID

Comment (andrewelamb) [reply tests/data/mock_manifests/filepath_submission_test_manifest.csv:1]: Yeah I'm worried this is going to break other tests...

Comment (jaymedina) [reply tests/data/mock_manifests/filepath_submission_test_manifest.csv:1]: Hi @andrewelamb - you raise a good point, however this test is the only one that references `filepath_submission_test_manifest.csv`, which is why it's best to change it now in case future integration tests reference it
image

This is what the search for that filename came up on GitHub

Comment (thomasyu888) [reply tests/data/mock_manifests/filepath_submission_test_manifest.csv:1]: All the tests seemed to pass via CI/CD, but can we triple check "Sample ID" is not referenced in the test.

Separately, if it was being referenced as `df._2`, were you seeing that in the tests?

Comment (andrewelamb) [reply tests/data/mock_manifests/filepath_submission_test_manifest.csv:1]: Awesome thanks for checking!

Comment (jaymedina) [reply tests/data/mock_manifests/filepath_submission_test_manifest.csv:1]: > can we triple check "Sample ID" is not referenced in the test?

I can confirm `Sample ID` is not referenced in this integration test in its current state.

>  if it was being referenced as df._2, were you seeing that in the tests?

You can view [this older commit ](https://github.com/Sage-Bionetworks/schematic/pull/1485/commits/48f10c2a69bd9a159e10ed8ad342b780262eb69d)which shows show the `Sample ID` column was referenced previously (line 60).

The `_2` is a byproduct of using `itertuples` on the dataframe. Here is a simple CSV file I made to recreate this:

image

Comment (andrewelamb) [reply tests/data/mock_manifests/filepath_submission_test_manifest.csv:1]: I want to get @linglp 's take on this before any approval. We allow users to have spaces in their manifest column names, so I want to make sure we aren't messing that up.

File: tests/integration/test_metadata_model.py
Comment (BryanFauble) [reply tests/integration/test_metadata_model.py]: No worries then! Thanks for trying it

**Review by BryanFauble — APPROVED** [review]
The logic is the changes looks great! I'll let fair folks give their final review though on the testing changes :D

File: tests/data/mock_manifests/filepath_submission_test_manifest.csv
Comment (jaymedina) [reply tests/data/mock_manifests/filepath_submission_test_manifest.csv:1]: Sounds good! Separately, I can add another integration test that covers for cases where the CSV files have column names with spaces.

**Comment by jaymedina** [comment]
There's a new test that's breaking due to a change on the Synapse end. The test that's breaking is [test_validate.py/test_filename_manifest](https://github.com/Sage-Bionetworks/schematic/blob/b038bf63db016f5ae2afd43fca22f1d1a748036c/tests/test_validation.py#L682)

Here is the expected error ([source](https://github.com/Sage-Bionetworks/schematic/blob/b038bf63db016f5ae2afd43fca22f1d1a748036c/tests/test_validation.py#L716)):

```
assert (
            GenerateError.generate_filename_error(
                val_rule="filenameExists",
                attribute_name="Filename",
                row_num="4",
                invalid_entry="schematic - main/MockFilenameComponent/txt5.txt",
                error_type="path does not exist",
                dmge=dmge,
            )[0]
            in errors
        )
```

Here is the actual error I'm getting:
```
In [7]: errors
Out[7]: 
[['3',
  'Filename',
  "The entityId for file path 'schematic - main/MockFilenameComponent/txt4.txt' on row 3 does not match the entityId for the file in the file view",
  'schematic - main/MockFilenameComponent/txt4.txt'],
 ['4',
  'Filename',
  "The entityId for file path 'schematic - main/MockFilenameComponent/txt5.txt' on row 4 does not match the entityId for the file in the file view",
  'schematic - main/MockFilenameComponent/txt5.txt']]
  ```

The file was added yesterday afternoon (my time) so the timing checks out with when this test started breaking ([link](https://www.synapse.org/Synapse:syn61682648)):

image

I can patch this up, but will do so on another PR to avoid scope creep on this one.

We might want to create standalone projects for some of these major test areas to avoid unexpected changes on Synapse. Or add logic to `helpers` that tears down and rebuilds for every test (or chunk of tests) so we have a blank slate each time. @BryanFauble I remember you started talking about this on Slack, what are your thoughts on this?

**Comment by andrewelamb** [comment]
@jaymedina I was just looking at this myself

**Comment by andrewelamb** [comment]
@jaymedina No I haven't I think you're jsut a bit ahead of me :) I don't quite understand what's going on here, could you elaborate?

**Comment by jaymedina** [comment]
It looks like this test was written back when `txt5.txt` in this [dataset_scope](https://www.synapse.org/Synapse:syn61682648) did not exist, so one of the expected errors from `validateModelManifest` is a `"path does not exist"` error type. However, because this file exists as of yesterday, I am not getting this error when I re-create the logic of this test. Rather, I'm getting another mismatched entity error:

image

**Comment by thomasyu888** [comment]
Cc @BWMac on this, as I think he might be working on this logic related to filename/entity ids

**Comment by andrewelamb** [comment]
So to summarize, the txt5.txt file was added by you yesterday, but there was a test that was specifically checking for the lack of a txt5.txt file, is that correct?

**Comment by andrewelamb** [comment]
> My understanding of how these errors are generated is limited, but I know changing the expected error for `txt5.txt` to a `mismatched entityId` fixes the problem:

I'd rather not chnage this test. @BWMac Can you work around this test as-is for your work on filename/entity ids?

**Comment by jaymedina** [comment]
Another suggestion that I think would work is to keep the `txt5.txt` file, but name it something else.

**Comment by thomasyu888** [comment]
> Another suggestion that I think would work is to keep the `txt5.txt` file, but name it something else. But I realize these are band-aid fixes to the real problem which is tests communicating with the same entities on Synapse which are subject to change

See this: https://sagebionetworks.slack.com/archives/C07HLTBS5D2/p1726156665980609?thread_ts=1726000362.049489&cid=C07HLTBS5D2

File: tests/data/mock_manifests/filepath_submission_test_manifest.csv
Comment (linglp) [reply tests/data/mock_manifests/filepath_submission_test_manifest.csv:1]: Apologies for the delayed response; I was out sick. we can't change the column name from Sample ID to SampleID. From what I understand, this manifest was generated using the example data model with the BulkRNA-seqAssay component data type. When I ran the manifest/generate function to create a manifest, the column headers I saw included "Sample ID" with the space. In this case, our test manifest should align with the output from manifest/generate, we need to keep the column name as is.

The issue that you refer @jaymedina , is because of this line: 
```
        manifest = helpers.get_data_frame(
            manifest_path, preserve_raw_input=False, **load_args
        )
```
If you change preserve_raw_input to True, and you also change how manifest was looped: 
```
for index, row in manifest.iterrows():
```
Does that solve the issue?

**Review by linglp — CHANGES_REQUESTED** [review]

File: tests/data/mock_manifests/filepath_submission_test_manifest_sampleidx10.csv
Comment (linglp) [inline tests/data/mock_manifests/filepath_submission_test_manifest_sampleidx10.csv:1]: @jaymedina , similarly here, since the test manifest is using component `BulkRNA-seqAssay`, the column name should be kept as "Sample ID"

File: tests/data/mock_manifests/ValidFilenameManifest.csv
Comment (linglp) [inline tests/data/mock_manifests/ValidFilenameManifest.csv:1]: here, since the component is `MockFilename`, I am fine with this column be kept as "SampleID".

File: tests/integration/test_metadata_model.py
Comment (linglp) [inline tests/integration/test_metadata_model.py:35]: here i am seeing three test cases here. Could you add some comments to explain what each test case is meant to test?

**Comment by BryanFauble** [comment]
> We might want to create standalone projects for some of these major test areas to avoid unexpected changes on Synapse. Or add logic to helpers that tears down and rebuilds for every test (or chunk of tests) so we have a blank slate each time. @BryanFauble I remember you started talking about this on Slack, what are your thoughts on this?

I would like us to eventually pivot to use the dev instance of Synapse to table all of these tests. At that point in time we can separate out all of the tests to prevent the usage of shared resources. In theory we should allow for each and every test to run at the exact same moment in time, however, I am aware there are many race conditions present in the code that modify shared resources - So it's not exactly realistic right now.

@linglp @andrewelamb I created this ticket to track this - Probably not high priority, but something to keep in mind:
https://sagebionetworks.jira.com/browse/FDS-2414

File: tests/integration/test_metadata_model.py
Comment (linglp) [inline tests/integration/test_metadata_model.py:90]: I am only seeing `manifest_record_type = "files_and_entities"` got tested. Can we also have `manifest_record_type="table_and_file"` got tested somewhere too? 

when submitting a manifest with parameter`manifest_record_type="table_and_file"`, this function `upload_manifest_as_table` (see [here](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/store/synapse.py#L1946)) would get triggered. And this function would do the following in terms of adding annotations: 
* adding annotations to entities listed in the manifest (see [here](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/store/synapse.py#L1997C22-L1997C55)) 
* setting annotations for the file itself (see [here](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/store/synapse.py#L2017)) 
* setting annotations for the table (see [here](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/store/synapse.py#L2017))

From my perspective, ideally the test case should test that annotations are added for all these three cases. Can we have this added in the PR, if not, can we open a new ticket? 

Also, when `manifest_record_type = "files_and_entities"`, when submitting a manifest, this function [here](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/store/synapse.py#L2042) got called. And this function would do the following in terms of adding annotations: 
* adding annotations to entities listed in the manifest (see [here](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/store/synapse.py#L2073))
* setting annotations for the file itself (see [here](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/store/synapse.py#L2093))

By reading your test, I am confident that bullet point 1 is tested, but bullet point 2 is not. Can we add test for bullet point 2?

**Comment by linglp** [comment]
@BWMac could you remove the txt5.txt here: https://www.synapse.org/Synapse:syn63106789
This is causing the test in develop branch to fail too: 
```
=============================================================== short test summary info ===============================================================
FAILED tests/test_validation.py::TestManifestValidation::test_filename_manifest - assert ['4', 'Filename', "The file path 'schematic - main/MockFilenameComponent/txt5.txt' on row 4 does not exist in the file view.", 'schematic -...
```
If this doesn't get resolved, other PRs also can't merge to develop. If you need to experiment, could you use a separate project or dataset? 

cc @jaymedina , @andrewelamb

**Comment by BWMac** [comment]
Sorry for not chiming in sooner, my email notifications for this PR weren't working for some reason.

I have removed text5.txt, but I created the file to add new tests while continuing the naming convention for testing `filename_validation`, not simply to experiment.

File: tests/data/mock_manifests/filepath_submission_test_manifest.csv
Comment (jaymedina) [reply tests/data/mock_manifests/filepath_submission_test_manifest.csv]: Hi @linglp , yes switching to `iterrows()` and preserving the raw input solved this issue without needing to alter `Sample ID`, thanks!

File: tests/integration/test_metadata_model.py
Comment (linglp) [reply tests/integration/test_metadata_model.py]: Hi Jenny! Sorry for the late reply. I was out yesterday. Your understanding that each manifest file could only contain one component type is correct 👍 . 

For your second question, I just noticed that we don't have a unit test for it (which I just create a ticket for it: https://sagebionetworks.jira.com/browse/FDS-2421). From my understanding, this function `format_manifest_annotations` handles adding annotations for manifest file itself and also the table. Here's how it works: 
* when manifest_record_type = "file_only", The code (see the "if" statement there) in `format_manifest_annotation` makes sure that component in the manifest gets captured in the annotations. In this case, Component = `BulkRNA-seqAssay` got added as an annotation key pair. As you could see below, the manifest csv got annotated like below eventually: 
Screenshot 2024-09-17 at 3 33 10 PM


* when manifest_record_type = "table_and_file", we add annotations to table: 
Screenshot 2024-09-17 at 3 38 14 PM
But as you could see, it only contains etag and id

Comment (jaymedina) [reply tests/integration/test_metadata_model.py]: @linglp thank you for the breakdown, that makes sense. It looks like [the logic of adding the `Component` annotation](https://github.com/Sage-Bionetworks/schematic/blob/5721c77d35ce14b45c7bf768af133c6d81eb4abe/schematic/store/synapse.py#L1618) is only triggered under `is_file` and not `is_table`, which is why there's no `Component` annotation in the tables. 

What is the purpose of adding this `Component` annotation to the manifest file itself? My first assumption would be: because it makes it easier to see what model is being used to annotate the entities, but then that raises the question of: why are the tables not annotated the same way?

File: tests/integration/test_metadata_model.py
Comment (linglp) [inline tests/integration/test_metadata_model.py:131]: we probably shouldn't download them to a fixed directory because what if one workflow
creates this temp directory, and another concurrent workflow attempts to delete it?  What do you think @BryanFauble ?

File: tests/integration/test_metadata_model.py
Comment (linglp) [inline tests/integration/test_metadata_model.py:196]: Thanks for the update Jenny! I think here you make sure that the table being submitted has the same columns as the columns in the manifest. But my previous comment also mentioned about the annotations of the table: 

Screenshot 2024-09-17 at 4 09 21 PM

Do you have time to add a test here and make sure the following?  
* the table gets annotated with the correct entity id 
* etag is not empty

If not, no worry. We could open an additional ticket 👍🏼

**Review by linglp — APPROVED** [review]
This overall looks pretty good. I left some comments with some additional thoughts. I am going to approve it if the downloading manifest to a fixed directory part is good after confirming with Bryan.

File: tests/integration/test_metadata_model.py
Comment (jaymedina) [reply tests/integration/test_metadata_model.py:196]: [Added](https://github.com/Sage-Bionetworks/schematic/pull/1485/commits/f32e756e40efdb440ccefb6c8f5bf8f9b84c6e1c)!

File: tests/integration/test_metadata_model.py
Comment (jaymedina) [reply tests/integration/test_metadata_model.py:180]: Hi @linglp - sounds good! To clarify before I move forward: I generalized this logic to validate the annotations for the manifest file/table because [this docstring ](https://github.com/Sage-Bionetworks/schematic/blob/5721c77d35ce14b45c7bf768af133c6d81eb4abe/schematic/store/synapse.py#L1606) was implying that, in the future, the manifest files/tables will be annotated with more annotations beyond `Component`.

With that in mind, I'll be keeping that generalized logic so that developers don't have to change this test in the future, but I will patch the other checks for `eTag` and `entityId` so it's easier to read. Thanks!

Comment (linglp) [reply tests/integration/test_metadata_model.py:180]: Oh I see. Thanks for pointing that out. What you proposed makes sense. Thank you!

Comment (jaymedina) [reply tests/integration/test_metadata_model.py]: [Updated](https://github.com/Sage-Bionetworks/schematic/pull/1485/commits/e3da3d2cf25a6e31e1a17be2d91878d01aa70264)

File: tests/integration/test_metadata_model.py
Comment (jaymedina) [reply tests/integration/test_metadata_model.py]: This is a good point @linglp, I've re-worked this logic to use a native python package called `tempfile` instead which builds and tears down a unique directory each time: https://github.com/Sage-Bionetworks/schematic/pull/1485/commits/e3da3d2cf25a6e31e1a17be2d91878d01aa70264

Comment (linglp) [reply tests/integration/test_metadata_model.py]: This is looking very good Jenny! Thank you again! I appreciate a lot!

**Comment by jaymedina** [comment]
Hi @linglp!

After the merge of https://github.com/Sage-Bionetworks/schematic/pull/1478 I made a few final changes to the scripts in this branch that needed to be updated. Namely:
1. I needed to update `process_row_annotations(...)` to move Gianna's solution in that method
2. I updated the tests to `process_row_annotations`, which are in `test_store_synapse.py` to have annotation dictionaries with the same structure as the annotations retrieved from Synapse

The change to these dictionary structures in the tests are best explained by Gianna's comment in the description:
>Using the client annotations are formated:
>{"ID":[], <file annotations as key value pairs>}
>Using the API the annotations are formatted:
>{'annotations':{'id': value,... 'annotations': {<file annotations as key value pairs>}}}

Since we are now retrieving the annotations using the API, I formatted the testing annotations accordingly

3. Lastly, I added a check in the `test.yaml` that removes corrupt venvs to prevent this error from breaking the builds:

```
The virtual environment found in /home/runner/work/schematic/schematic/.venv seems to be broken.
Recreating virtualenv schematicpy in /home/runner/work/schematic/schematic/.venv
``` 
Whether or not this will stay in the PR is contingent on if we decide to keep the virtual environment cache-ing in `test.yml`.

----

I'm going to tag you in the spots where the changes were made so it's easier to sift through. Thanks again for your time

File: schematic/store/synapse.py
Comment (jaymedina) [inline schematic/store/synapse.py:1581]: Here is the start of the changes to `process_row_annotations` @linglp

File: tests/integration/test_store_synapse.py
Comment (jaymedina) [inline tests/integration/test_store_synapse.py:37]: @linglp

**Comment by linglp** [comment]
@jaymedina Thanks Jenny. It looks good to me. Could you highlight the changes of workflow in the PR description?

**Comment by jaymedina** [comment]
@linglp Sure thing - [Here](https://github.com/Sage-Bionetworks/schematic/pull/1485#issue-2495487646) is the new description. Thanks! I'll merge this in once the final integration test on py 3.10 passes

**Pull request merged**
